enterprise/middleware.py

```diff
@@ -0,0 +1,28 @@
+from enterprise.models import EnterpriseCustomer, UserDataSharingConsentAudit
+from enterprise.tpa_pipeline import (
+    active_provider_enforces_data_sharing,
+    get_consent_status_for_request,
+    get_enterprise_customer_for_request
+)
+class ResetSessionIfPipelineBrokenMiddleware(object):
+    """
+    Middleware signs the user out if they need to provide data sharing consent,
```

Comment:
So, I've made the middleware more generic, and moved it back into third_party_auth. However, based on my findings, a general solution for restricting the user to the TPA pipeline at large is not forthcoming.

The reason for this is that individual pipeline elements don't actually receive any request context to work with; they receive, essentially, an expansion of the pipeline's `kwargs` variable.

This is stored on the context, and is manipulable, but the problem is that it's only manipulable by actually _returning_ a dictionary with the variables to update. The problem being, of course, that the cases where we would need something like this most would be in cases where, rather than a dictionary that manipulates the pipeline kwargs, we're already returning a `redirect` object that sends the user to a difference page, and don't have the ability to return something else that would mutate the pipeline state.

Long story short, to have a fully-protected pipeline, we'd have to have something in the actual views that adds that protection, which is somewhat outside the scope here.